src/type_checker/checker.rs

```diff
@@ -610,6 +610,34 @@ impl<B: Backend> TypeChecker<B> {
                 // exit the scope
                 typed_fn_env.pop();
             }
+            StmtKind::IteratorLoop {
+                var,
+                iterator,
+                body,
+            } => {
+                // enter a new scope
+                typed_fn_env.nest();
+
+                // make sure that the iterator expression is an iterator,
+                // for now this means that the expression should have type `Array`
+                let iterator_typ = self.compute_type(iterator, typed_fn_env)?;
+                let iterator_typ =
+                    iterator_typ.expect("Could not compute type of iterator (TODO: better error)");
```

Comment:
this can be an error thrown with a span, so it can better inform the users where the error is originated in the code.